### PR TITLE
Add role based authorization

### PR DESCRIPTION
**Pull request by trueleo**
Part of #250

### CODE REVIEW

**Review by nitisht — CHANGES_REQUESTED** [review]

File: server/src/handlers/http.rs [hunk 1]
```diff
@@ -135,69 +138,111 @@ pub fn configure_routes(cfg: &mut web::ServiceConfig) {
...
+                .route(web::delete().to(rbac::delete_user).auth(Action::DeleteUser))
+                .route(web::put().to(rbac::put_roles).auth(Action::PutUser)),
+        )
+        .service(
+            web::resource("/{username}/roles")
+                // PUT /user/{username}/roles => Put roles for user
+                .route(web::put().to(rbac::put_roles).auth(Action::PutRoles)),
+        )
+        .wrap_fn(|req, srv| {
+            // The credentials set in the env vars (P_USERNAME & P_PASSWORD) are treated
```
Comment (nitisht) [inline server/src/handlers/http.rs:219]: Can we separate this in a separate private function - easier to read

File: server/src/rbac/role.rs [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+        RoleBuilder {
+            actions: vec![
+                Action::Query,
+                Action::GetSchema,
+                Action::GetStats,
+                Action::GetRetention,
+                Action::GetAlert,
+            ],
+            stream: None,
+            tag: None,
```
[the remark this answers is not in the document]
Comment (trueleo) [reply server/src/rbac/role.rs:156]: Yes

File: server/src/handlers/http.rs [hunk 3]
```diff
@@ -135,69 +138,111 @@ pub fn configure_routes(cfg: &mut web::ServiceConfig) {
         .service(
             web::resource("")
                 // PUT "/logstream/{logstream}" ==> Create log stream
-                .route(web::put().to(logstream::put_stream))
+                .route(
+                    web::put()
+                        .to(logstream::put_stream)
+                        .auth_stream(Action::CreateStream),
```
[the remark this answers is not in the document]
Comment (trueleo) [reply server/src/handlers/http.rs:144]: auth and auth_stream are extention function for middleware. `stream`here is for checking both action and stream for given user
 
```rust
pub struct Authorization {
    pub action: Action,
    pub stream: bool,
}
```

[the remark this answers is not in the document]
Comment (nitisht) [reply server/src/handlers/http.rs:144]: then a better name would be `authorise_for_stream`

**Comment by trueleo** [comment]
>   `{{endpoint}}/api/v1/logstream/{{stream_name}}` succeeds for a user with `reader` role.

Fixed

File: server/src/handlers/http.rs [hunk 4]
```diff
@@ -135,69 +138,111 @@ pub fn configure_routes(cfg: &mut web::ServiceConfig) {
...
+        .service(
+            web::resource("/{username}")
+                // PUT /user/{username} => Create a new user
+                .route(web::put().to(rbac::put_user).auth(Action::PutUser))
+                // DELETE /user/{username} => Delete a user
+                .route(web::delete().to(rbac::delete_user).auth(Action::DeleteUser))
+                .route(web::put().to(rbac::put_roles).auth(Action::PutUser)),
+        )
+        .service(
+            web::resource("/{username}/roles")
```
[the remark this answers is not in the document]
Comment (nitisht) [reply server/src/handlers/http.rs:222]: oh I meant `api/v1/user/{username}/permission`

File: server/src/handlers/http.rs (hunk 1 above)
Comment (trueleo) [reply server/src/handlers/http.rs:227]: Can change this to wrap .. but that will require more boilerplate for this middleware. ( similar to middleware defined in auth_middleware.rs

Comment (nitisht) [reply server/src/handlers/http.rs:227]: Still okay because it will make readability much better

**Review by nitisht — APPROVED** [review]
LGTM & Tested

**Pull request merged**